### PR TITLE
Userscope sourcecoderesource

### PR DESCRIPTION
**Pull request by gitlawr**
Dependent on https://github.com/rancher/types/pull/284
Related issue: https://github.com/rancher/rancher/issues/11977

1. change sourceCodeCredential/Repository to user-scoped( stored in user namespace) so users won't have access to other users account/repositories.
2. Because Pipelines are in project-scoped. When user A sees user B's pipelines we need some info of B's Github account to show(like avatar, Github username, etc.), so we store a sourcecodecredential reference in pipelinestatus with sensitive data erased.
3. Do some refactor on pipeline lifecycle meanwhile.

### CODE REVIEW

File: pkg/api/customization/pipeline/source_code_credential.go [hunk 1]
```diff
@@ -63,8 +64,13 @@ func (h *SourceCodeCredentialHandler) ActionHandler(actionName string, action *t
...
 
 func (h *SourceCodeCredentialHandler) refreshrepos(apiContext *types.APIContext) error {
-
-	credential, err := h.SourceCodeCredentialLister.Get("", apiContext.ID)
+	parts := strings.Split(apiContext.ID, ":")
+	if len(parts) <= 1 {
+		return httperror.NewAPIError(httperror.InvalidFormat, "invalid Id")
+	}
+	ns := parts[0]
+	name := parts[1]
```
Comment (zionwu) [inline pkg/api/customization/pipeline/source_code_credential.go]: we can use the API in pkg/ref:
```
ns, name :=ref.Parse(apiContext.ID)
```

File: pkg/api/customization/pipeline/utils.go [hunk 2]
```diff
@@ -11,6 +12,8 @@ import (
 func refreshReposByCredential(sourceCodeRepositories v3.SourceCodeRepositoryInterface, sourceCodeRepositoryLister v3.SourceCodeRepositoryLister, credential *v3.SourceCodeCredential) ([]*v3.SourceCodeRepository, error) {
 
 	remoteType := credential.Spec.SourceCodeType
+	nameSpace := credential.Namespace
```
Comment (zionwu) [inline pkg/api/customization/pipeline/utils.go]: `namespace` would be better?

File: pkg/controllers/user/pipeline/engine/jenkins/engine.go [hunk 3]
```diff
@@ -443,7 +443,13 @@ func (j Engine) setCredential(pipeline *v3.Pipeline) error {
 		return errors.New("Invalid pipeline definition")
 	}
 	credentialID := pipeline.Spec.Stages[0].Steps[0].SourceCodeConfig.SourceCodeCredentialName
-	souceCodeCredential, err := j.SourceCodeCredentialLister.Get("", credentialID)
+	parts := strings.Split(credentialID, ":")
+	if len(parts) <= 1 {
+		return fmt.Errorf("error invalid credential Id '%s'", credentialID)
+	}
+	ns := parts[0]
+	name := parts[1]
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/engine/jenkins/engine.go]: use ref.Parse here

File: pkg/controllers/user/pipeline/controller/pipeline/pipeline.go [hunk 4]
```diff
@@ -47,27 +49,53 @@ func Register(ctx context.Context, cluster *config.UserContext) {
...
+	if !utils.ValidSourceCodeConfig(obj.Spec) {
+		return obj, fmt.Errorf("error invalid definition of pipeline '%s'", obj.Spec.DisplayName)
+	}
+	sourceCodeCredentialID := obj.Spec.Stages[0].Steps[0].SourceCodeConfig.SourceCodeCredentialName
+	parts := strings.Split(sourceCodeCredentialID, ":")
+	if len(parts) <= 1 {
+		return obj, fmt.Errorf("error invalid sourceCodeCredentialId in pipeline '%s'", obj.Spec.DisplayName)
+	}
+	ns := parts[0]
+	name := parts[1]
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: ref.Parse

File: pkg/controllers/user/pipeline/controller/pipeline/pipeline.go [hunk 5]
```diff
@@ -101,24 +129,18 @@ func (l *Lifecycle) Updated(obj *v3.Pipeline) (*v3.Pipeline, error) {
...
-	credentialName := obj.Spec.Stages[0].Steps[0].SourceCodeConfig.SourceCodeCredentialName
-	credential, err := l.sourceCodeCredentialLister.Get("", credentialName)
+	credentialID := obj.Spec.Stages[0].Steps[0].SourceCodeConfig.SourceCodeCredentialName
+	parts := strings.Split(credentialID, ":")
+	if len(parts) <= 1 {
+		return "", fmt.Errorf("error invalid credential Id '%s'", credentialID)
+	}
+	ns := parts[0]
+	name := parts[1]
+	credential, err := l.sourceCodeCredentialLister.Get(ns, name)
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: ref.Parse

File: pkg/controllers/user/pipeline/controller/pipeline/pipeline.go [hunk 6]
```diff
@@ -145,8 +167,14 @@ func (l *Lifecycle) deleteHook(obj *v3.Pipeline) error {
...
 	}
-	credentialName := obj.Spec.Stages[0].Steps[0].SourceCodeConfig.SourceCodeCredentialName
-	credential, err := l.sourceCodeCredentialLister.Get("", credentialName)
+	credentialID := obj.Spec.Stages[0].Steps[0].SourceCodeConfig.SourceCodeCredentialName
+	parts := strings.Split(credentialID, ":")
+	if len(parts) <= 1 {
+		return fmt.Errorf("error invalid credential Id '%s'", credentialID)
+	}
+	ns := parts[0]
+	name := parts[1]
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: ref.Parse

**Comment by gitlawr** [comment]
Thanks for the review, I use the `pkg/ref` to clean codes in some other places as well.

**Comment by zionwu** [comment]
LGTM

File: pkg/controllers/user/pipeline/controller/pipeline/pipeline.go [hunk 7]
```diff
@@ -47,27 +49,49 @@ func Register(ctx context.Context, cluster *config.UserContext) {
...
+
+	return l.sync(obj)
+}
+
+func (l *Lifecycle) Remove(obj *v3.Pipeline) (*v3.Pipeline, error) {
+
+	if obj.Status.WebHookID != "" {
+		if err := l.deleteHook(obj); err != nil {
+			//merely log error to avoid deletion block
+			logrus.Warningf("fail to delete previous set webhook for pipeline '%s' - %v", obj.Spec.DisplayName, err)
```
Comment (alena1108) [inline pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: Nit: use Warnf. Also is situation when previous webhook deletion failed, recoverable? What are the implication of the failed deletion?

Comment (gitlawr) [reply pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: I may not be recoverable. For example when the user already deleted the webhook in this repo manually. So if the deletion failed we delete we proceed anyway.

**Pull request merged**